csharp/src/Drivers/Apache/Hive2/HiveServer2Statement.cs

```diff
@@ -587,5 +592,370 @@ protected internal QueryResult EnhanceGetColumnsResult(Schema originalSchema, IR
+                        case ArrowTypeId.Int16:
+                            ((Int16Array.Builder)_builder).Append(((Int16Array)columnArray).GetValue(rowIndex)!.Value);
+                            break;
+                        case ArrowTypeId.Int32:
+                            ((Int32Array.Builder)_builder).Append(((Int32Array)columnArray).GetValue(rowIndex)!.Value);
+                            break;
+                        default: // Handles String and other types that default to StringArray.Builder
+                            // Try to cast to StringArray and get string value - if it fails, 
+                            // the outer try-catch will handle it and call AppendNull()
+                            ((StringArray.Builder)_builder).Append(((StringArray)columnArray).GetString(rowIndex));
+                            break;
+                    }
+                }
+                catch
+                {
+                    // If any conversion fails or columnArray is null, append null as fallback
+                    AppendNull();
+                }
+            }
+            
+     
+        }
+
+        private async Task<QueryResult> GetColumnsExtendedAsync(CancellationToken cancellationToken = default)
+        {
+            // 1. Get all three results at once
+            var columnsResult = await GetColumnsAsync(cancellationToken);
+            if (columnsResult.Stream == null) return columnsResult;
+
+            var pkResult = await GetPrimaryKeysAsync(cancellationToken);
+
+            // For FK lookup, we need to pass in the current catalog/schema/table as the foreign table
+            var resp = await Connection.GetCrossReferenceAsync(
+                null,
+                null,
+                null,
+                CatalogName,
+                SchemaName,
+                TableName,
+                cancellationToken);
+
+            var fkResult = await GetQueryResult(resp.DirectResults, cancellationToken);
+
+            // 2. Read all batches into memory
+            List<RecordBatch> columnsBatches;
+            int totalRows;
+            Schema columnsSchema;
+            StringArray? columnNames = null;
+            int colNameIndex = -1;
+
+            // Extract column data
+            using (var stream = columnsResult.Stream)
+            {
+                colNameIndex = stream.Schema.GetFieldIndex("COLUMN_NAME");
+                if (colNameIndex < 0) return columnsResult; // Can't match without column names
+                
+                var batchResult = await ReadAllBatchesAsync(stream, cancellationToken);
+                columnsBatches = batchResult.Batches;
+                columnsSchema = batchResult.Schema;
+                totalRows = batchResult.TotalRows;
+                
+                if (columnsBatches.Count == 0) return columnsResult;
+                
+                // Create column names array from all batches
+                var builder = new StringArray.Builder();
```

Comment:
For what it's worth, there's an `ArrayDataConcatenator` in the base Arrow libraries that might be useful here -- in addition to being considerably more efficient. It's used in `ListArrayExtensions` in this ADBC codebase.